---
title: "SCS Analysis"
author: "Aubrey Odom"
date: "2024-07-08"
output: html_document
---
# Initial exploration

- Looked for all files in folders of interest:
ls -LR outs/ > all_outs_files.txt

/projectnb/altcells/singlecellseq/Data/Combined_2023-06-06_2023-08-15_ChitaliaV/JA_CV_0522_4_control/all_outs_files.txt

## Identifying the various outputs.

web_summary.html
metrics_summary.csv
Gives an overview of the number of cells, mean reads, median genes per cell (10X genomics CellRanger output)

raw_feature_bc_matrix
raw_feature_bc_matrix.h5

filtered_feature_bc_matrix
filtered_feature_bc_matrix.h5

cloupe.cloupe
From cell ranger, "The .cloupe file can then be imported into Loupe Browser v7.0 for data visualization and further exploration."

molecule_info.h5

# Evan's instructions for running SCS analysis
- Art and Xutao are single cell experienced users

1) Run Cell Ranger (already done by SC core)
2) Utilize the Seurat vignette to visualize the data
3) Identify the cells
4) Find the interesting cells
5) Dissect the cluster dynamics
6) Perform differential expression between those two clusters,
		○ So take the members of the clusters
		○ What genes are differentially expressed

# Setup
```{R}
suppressPackageStartupMessages({
  library(tidyverse)
  library(dplyr)
  library(Seurat)
  library(patchwork)
  library(ggplot2)
  # /satijalab/seurat-data
})

```

## Import data
The four groups...
"PDtreatment" 5 mg/kg PTX + 5 mg/kg DEX 
"Ptreatment" 5 mg/kg PTX
"Dtreatment" 5 mg/kg DEX
"control" 5% DMSO CTL


ORIGINAL READ COUNTS OF CELLS (from web_summary.html)

PTreatment - 20,791
PDTreatment - 4,243
DTreatment - 6,011
control - 4,996
```{R}
stem <- "/projectnb/altcells/singlecellseq/Data/Combined_2023-06-06_2023-08-15_ChitaliaV"
conditions <- list.files(stem)
cond_names <- sapply(conditions, function(x) stringr::str_split_i(x, "_", 5))
names(conditions) <- cond_names
path_to_matrix <- "outs/filtered_feature_bc_matrix"

this_folder <- conditions[1]

read_in_data <- function(this_folder) {
  path_to_data <- file.path(stem, this_folder, path_to_matrix)
  
  raw_counts <- Read10X(path_to_data, gene.column = 2, cell.column = 1,
                        unique.features = TRUE, strip.suffix = FALSE)
  # Quality CONTROL - downsample to min(5000, num)
  down_num <- min(5000, ncol(raw_counts))
  ds_ind <- sample(size = down_num, x = ncol(raw_counts), replace = FALSE)
  # Samples on the columns
  raw_counts_ds <- raw_counts[, ds_ind]
  this_trt <- stringr::str_split_i(this_folder, "_", 5)
  meta_data <- data.frame(Treatment = rep(this_trt, ncol(raw_counts_ds)), 
                          row.names = colnames(raw_counts_ds))
  sobject <- CreateSeuratObject(counts = raw_counts_ds, min.cells = 0, min.features = 0,
                                project = this_trt,
                                meta.data = meta_data) |>
    SCTransform(vst.flavor = "v2", verbose = FALSE) |>
    # Using assay from SCTransform function
    FindVariableFeatures(selection.method = "vst", nfeatures = 2000, assay = "SCT")
}

data_list <- plyr::alply(conditions, 1, read_in_data, .progress = "text",
                         .dims = TRUE)

features <- SelectIntegrationFeatures(object.list = data_list)
```
# Analysis 

## Perform integration
```{R}
data_anchors <- data_list |>
  PrepSCTIntegration(assay = "SCT") |>
  FindIntegrationAnchors(anchor.features = features, normalization.method = "SCT")
data_combined_pre <- IntegrateData(anchorset = data_anchors,
                                   normalization.method = "SCT")
```

## Perform PCA
```{R}
DefaultAssay(data_combined_pre) <- "SCT"
data_combined <- ScaleData(data_combined_pre, verbose = FALSE) %>%
  RunPCA(npcs = 200, verbose = FALSE)
```

## Export data
```{R}
saveRDS(data_combined,
        file = "~/altcells/singlecellseq/vitalia_scs_aro/processed_data/data_combined.RDS")
```
