---
title: "05_Visualizations"
author: "Aubrey Odom"
date: "2024-08-28"
output: html_document
---

# Setup
```{R}
suppressPackageStartupMessages({
  library(tidyverse)
  library(dplyr)
  library(Seurat)
  library(SingleR)
  library(celldex)
  library(BiocParallel)
  library(SingleCellExperiment)
  library(randomcoloR)
  library(limma)
  library(ggiraphExtra)
  library(randomcoloR)
  library(ggrepel)
  # /satijalab/seurat-data
})

fig_folder <- "~/altcells/singlecellseq/vitalia_scs_aro/figures_tm"
```

# Read in data
```{R}
data_reclustered <- readRDS("~/altcells/singlecellseq/vitalia_scs_aro/processed_data/data_reclustered.RDS")

```

# Average read count among cells
```{R}
RNA_df <- tibble(RNA_count = data_reclustered$nCount_RNA,
                 TRT = data_reclustered$Treatment) 

# Among all cells
RNA_df |>
  summarise(meanRNA = mean(RNA_count))

# By group
RNA_df |>
  group_by(TRT) |>
  summarise(meanRNA = mean(RNA_count),
            medianRNA = median(RNA_count))

# Sig diff between groups??
anova_result <- aov(RNA_count ~ TRT, data = RNA_df)
summary(anova_result)

# YES
```

# Most abundant cell types
```{R}
table(data_reclustered$SingleR.pruned) |>
  sort()

```

## Obtain cell type distributions
```{R}
data_reclustered@meta.data %>%
  dplyr::count(SingleR.pruned) %>%
  mutate(percentage = 100 * n / sum(n),
         percentage = round(percentage, 2)) |>
  arrange(desc(percentage))

```

# UMAP

### UMAP with cell type + cluster
```{R}
DimPlot(data_reclustered, reduction = "umap", label = "TRUE",
        alpha = 0.5, group.by = "clusters") +
  ggplot2::labs(x = "UMAP Axis 1", y = "UMAP Axis 2",
                title = "Combined UMAP Plot",
                subtitle = "Clusters across treatments") +
  theme_bw()  +
  paletteer::scale_color_paletteer_d("tvthemes::gravityFalls",
                                     direction = 1)


ggsave(file.path(fig_folder, "UMAP_re_clustered.png"), width = 8, height = 5)
```

### UMAP with cell type only (if cell types are same, ignored cluster)

```{R}
# For colors
all_types <- unique(data_reclustered@meta.data$celltype)
set.seed(51)
palette <- distinctColorPalette(length(all_types)) |>
  setNames(all_types)

DimPlot(data_reclustered, reduction = "umap", label = "TRUE",
        alpha = 0.5, group.by = "celltype", cols = palette,
        label.box = FALSE, repel = TRUE) +
  ggplot2::labs(x = "UMAP Axis 1", y = "UMAP Axis 2",
                title = "UMAP of aorta cell types",
                subtitle = "Across all treatments") +
  theme_bw()

ggsave(file.path(fig_folder, "UMAP_celltype_only.png"), width = 8, height = 5)
```

### UMAP without clustering at all
```{R}
# For colors
all_types <- unique(data_reclustered@meta.data$SingleR.pruned)
set.seed(3)
palette <- distinctColorPalette(length(all_types)) |>
  setNames(all_types)

DimPlot(data_reclustered, reduction = "umap", label = "TRUE",
        alpha = 0.5, group.by = "SingleR.pruned", cols = palette,
        label.box = FALSE, repel = TRUE) +
  ggplot2::labs(x = "UMAP Axis 1", y = "UMAP Axis 2",
                title = "UMAP of aorta cell types",
                subtitle = "Across all treatments (sans clustering") +
  theme_bw()

ggsave(file.path(fig_folder, "UMAP_celltype_only_no_clusters.png"), width = 8, height = 5)
```

# Visualize strongest marker expression

### most variable genes

```{R}
all_rec_markers <- readRDS(file = "~/altcells/singlecellseq/vitalia_scs_aro/processed_data/all_markers_TM.RDS")
var_table_rec <- all_rec_markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1) %>%
  arrange(desc(avg_log2FC))

table(var_table_rec$cluster)
var_table_rec
```

```{R}
b_cell_markers <- read.csv("~/altcells/singlecellseq/vitalia_scs_aro/markers.csv")

all_markers <- all_rec_markers %>%
  filter(cluster == 10) %>%
  mutate(big_change = abs(avg_log2FC)) %>%
  arrange(desc(big_change)) %>%
  mutate()
  #slice_head(n = 20) %>%
  #View()

comb <- c(b_cell_markers$Synonyms, b_cell_markers$Synonyms)


matches <- sapply(comb[comb != ""], function(input) agrep(
  pattern = input, x = all_markers$gene,
                                  max.distance = 0.05,
                                  value = TRUE))
```


### Violin plots
```{R}
create_vlnplot <- function(gene_in) {
  p <- VlnPlot(data_reclustered, features = gene_in) +
  labs(title = "Inter-cluster differential genes",
       subtitle = gene_in) +
  xlab("Cluster Number") +
  theme_bw() +
  theme(legend.position = "none")
  return(p)
}

plots <- lapply(var_table_rec$gene[1:10], create_vlnplot) |>
  gridExtra::marrangeGrob(nrow = 4, ncol = 3)

plots

ggsave(plot = plots, file.path(fig_folder, "violinplot.png"),
       width = 10,
       height = 10)
```

### Feature plots

```{R}
FeaturePlot(data_reclustered, features = var_table_rec$gene[1:10],
            order = TRUE)

ggsave(file.path(fig_folder, "FeaturePlot.png"),
       width = 15,
       height = 10)
```

# Differentially Expressed Genes (DEG)

## Top genes per cluster

```{R}
top_genes_per_cluster <- var_table_rec %>%
  group_by(cluster) %>%
  arrange(desc(avg_log2FC)) %>%
  slice_head(n = 20) %>%
  relocate(cluster, gene, avg_log2FC) %>%
  mutate(avg_log2FC = round(avg_log2FC, 2),
         p_val_adj = round(p_val_adj, 2),
         p_val = round(p_val, 2))

write.csv(top_genes_per_cluster,
          file.path(fig_folder, "top_20_genes_per_cluster.csv"))
```

## Using Limma

### Inter-treatment DEG analysis

```{R}
# get names of diff exp genes
to_sub <- rownames(SeuratObject::GetAssayData(data_reclustered))

mat <- SeuratObject::GetAssayData(data_reclustered)[to_sub, ]

for(trt2 in c("PDtreatment", "Dtreatment", "control")) {
  trt1 <- "Ptreatment"
   # Design matrix
  model_data <- data_reclustered@meta.data |>
    dplyr::select(Treatment, SingleR.pruned, clusters) |>
    #mutate(SingleR.pruned = stringr::str_replace_all(SingleR.pruned, " ", "_"),
    #       group = paste(Treatment, SingleR.pruned, sep = "_")) |>
    filter(Treatment %in% c(trt1, trt2))
  design_matrix <- model.matrix(~0 + Treatment, data = model_data)
  
  # TMM normalization
  ind <- colnames(mat) %in% rownames(design_matrix)
  #dge <- mat[, ind ]  |>
  #  DGEList() |> calcNormFactors(method = "TMM")
  #
  ## Apply voom
  #v1 <- voom(dge, design_matrix, plot = TRUE)
  
  # Fit lm
  fit1 <- lmFit(mat[, ind ], design_matrix)
  
  # Make contrasts matrix
  these_contrasts <- makeContrasts(contrast = paste(paste0("Treatment", trt1), "-", paste0("Treatment", trt2)),
                                   levels = design_matrix)
  message("contrasts created")
  # Perform analysis
  fit2 <- contrasts.fit(fit1, these_contrasts) |> eBayes()
  
  deg_results <- topTable(fit2, adjust.method = "fdr",
                          sort.by = "B") |>
    dplyr::filter(`adj.P.Val` < 0.05)
  
  which_trt <- stringr::str_remove_all(c(trt1, trt2), "treatment")
  write.csv(deg_results, file.path(fig_folder, paste0("top_DEG_", paste(which_trt, collapse = "_"), ".csv")))
}

rm(mat)
```


# Radar Charts

This plot will contain the top 10 differentially expressed genes between control/paclitaxel, with four facets (one for each treatment)

## DEGs between Paclitaxel and Control
```{R}
markers_p_c <- read_csv(file.path(fig_folder, "top_DEG_P_control.csv")) |>
  dplyr::rename("gene" = `...1`)

n_top_genes <- 10
genes_of_interest <- markers_p_c %>%
  mutate(fc_strength = abs(logFC)) %>%
  arrange(desc(fc_strength))
```

```{R}
your_genes <- genes_of_interest |>
  filter(logFC > 0) |>
  slice_head(n = n_top_genes) |>
  pull(gene)

build_chart <- function(your_genes, lab) {
  avg_expr <- AverageExpression(data_reclustered, features = your_genes, group.by = c("seurat_clusters", "Treatment"),
                                assays = "integrated", slot = "scale.data",
                                normalization.method = "integrated", scale.factor = 100, layer = "scale.data")$integrated
  
  radar_dat <- avg_expr %>%
    as.matrix() %>%
    as.data.frame() %>%
    rownames_to_column("Gene") %>%
    pivot_longer(cols = -1,
                 names_to = c("Cluster", "Treatment"),
                 names_sep = "_",
                 values_to = "value") %>%
    mutate(Cluster = as.numeric(stringr::str_remove_all(Cluster, "g"))) %>%
    pivot_wider(names_from = Cluster) %>%
    mutate(Treatment = recode(Treatment,
                              "control" = "Control",
                              "Dtreatment" = "Dexmethasone",
                              "PDtreatment" = "Paclitaxel + Dexmethasone",
                              "Ptreatment" = "Paclitaxel"))
  
  set.seed(51)
  palette <- distinctColorPalette(n_top_genes)
  
  input_dat <- radar_dat %>%
    group_by(Treatment) %>%
    group_split() 
  
  mk_radar_plot <- function(data) {
    this_trt <- unique(data$Treatment)
    data |>
      ggRadar(aes(group = "Gene"), alpha = 0.3) + # alpha for polygon colors intensity
      theme_bw() +
      ggplot2::scale_fill_manual(values = palette) +
      ggplot2::scale_color_manual(values = palette) +
      labs(title = paste("Top", n_top_genes, "enriched genes between Paclitaxel and Control"),
           subtitle = this_trt, caption = lab) +
      theme(axis.text = element_text(size = 7))
    ggsave(file.path(fig_folder, "Radar", paste0(lab, "Radarchart_", stringr::str_remove(this_trt, " \\+ "), ".png")),
           units = "in", width = 7, height = 5)
  }
  
  lapply(input_dat, mk_radar_plot)
}

genes_of_interest |>
  filter(logFC > 0) |>
  slice_head(n = n_top_genes) |>
  pull(gene) |>
  build_chart(lab = "Upregulated")

genes_of_interest |>
  filter(logFC < 0) |>
  slice_head(n = n_top_genes) |>
  pull(gene) |>
  build_chart(lab = "Downregulated")

genes_of_interest |>
  slice_head(n = n_top_genes) |>
  pull(gene) |>
  build_chart(lab = "Most differentially expressed")
```

# Cell type UMAP

```{R}
# For colors
all_types <- unique(data_reclustered@meta.data$SingleR.pruned)
set.seed(51)
palette <- distinctColorPalette(length(all_types)) |>
  setNames(all_types)

DimPlot(data_reclustered, reduction = "umap", label = "TRUE",
        alpha = 0.5, group.by = "SingleR.pruned", cols = palette,
        label.box = FALSE, repel = TRUE) +
  ggplot2::labs(x = "UMAP Axis 1", y = "UMAP Axis 2",
                title = "UMAP of aorta cell types",
                subtitle = "Across all treatments") +
  theme_bw()

ggsave(file.path(fig_folder, "UMAP_celltype_only.png"), width = 8, height = 5)

```


# Cluster/treatment count

```{R}
all_counts <- tibble(cluster = Idents(data_reclustered), trt = data_reclustered$Treatment) |>
  arrange(trt) |>
  group_by(trt, cluster) |>
  summarise(count = n())

pivot_wider(all_counts, id_cols = c("cluster"), names_from = "trt", values_from = "count")

friedman.test(all_counts$count, groups = all_counts$trt, blocks = all_counts$cluster)

all_counts |>
  ggplot(aes(x = cluster, y = count, color = trt)) +
  geom_point() +
  theme_grey() +
  labs(title = "Cluster cell count by treatment") +
  scale_color_manual(values = c("royalblue", "brown", "grey4", "green4") %>%
                       setNames(c("Dtreatment", "Ptreatment", "PDtreatment", "control")))

ggsave(file.path(fig_folder,
                 paste0("Cluster_cell_count.png")),
       width = 5, height = 3)
```

# Stacked bar plots
```{R}
order_celltypes <- data_reclustered@meta.data |>
  group_by(SingleR.pruned) |>
  summarise(count = n()) |>
  arrange((count)) |>
  pull(SingleR.pruned)

plotting_data <- data_reclustered@meta.data |>
  mutate("Cell type" = factor(SingleR.pruned, levels = order_celltypes))

set.seed(50)
these_names <-unique(plotting_data$SingleR.pruned)
n <- length(these_names)
palette <- distinctColorPalette(n)
names(palette) <- these_names

```


## All clusters in one grid
```{R}

plotting_data |>
    ggplot2::ggplot(aes(x = Treatment, fill = `Cell type`)) +
    geom_bar(stat = "count") +
    ggplot2::scale_fill_manual(values = palette) +
    theme_bw() +
    labs(main = "Proportion of finer cell types in a given treatment") +
    ggplot2::xlab("Treatment") +
    ggplot2::ylab("Count") +
    ggplot2::coord_flip() +
    ggplot2::facet_wrap(~clusters) +
    ggplot2::labs(title = "Proportion of cell type by treatment and cluster")
  
  ggsave(file.path(fig_folder, "celltypes_cluster_trt.png"), height = 5, width = 8, units = "in")
  
```

## Individual plots
```{R}
for (clust in unique(plotting_data$clusters)) {
  plotting_data |>
    filter(clusters == clust) |>
    group_by(Treatment, `Cell type`) |>
    summarise(count = n(), .groups = "drop") |>
    mutate(prop = paste0(round(count / sum(count), 2)*100, "%")) |>  # Proportion of each group in its category
    ggplot2::ggplot(aes(x = Treatment, y = count, fill = `Cell type`, label = prop)) +
    geom_bar(stat = "identity") +
    geom_text(aes(label = prop), position = position_stack(reverse = FALSE, vjust = 0.5),
              size = 2) +
    ggplot2::scale_fill_manual(values = palette) +
    theme_bw() +
    labs(main = "Proportion of finer cell types in a given treatment") +
    ggplot2::xlab("Treatment") +
    ggplot2::ylab("Count") +
    ggplot2::coord_flip() +
    ggplot2::labs(title = "Proportion of cell type by treatment and cluster",
                  subtitle = paste0("Cluster ", clust))
  
  ggsave(file.path(fig_folder, "cluster_plots", 
                   paste0("celltypes_cluster", clust, "_trt.png")), height = 3.5, width = 7, units = "in")
}
```


# Radar Charts

This plot will contain the top 10 differentially expressed genes between control/paclitaxel, with four facets (one for each treatment)

## DEGs between Paclitaxel and Control
```{R}
markers_p_c <- read_csv(file.path(fig_folder, "top_DEG_P_control.csv")) |>
  dplyr::rename("gene" = `...1`)

n_top_genes <- 10
genes_of_interest <- markers_p_c %>%
  mutate(fc_strength = abs(logFC)) %>%
  arrange(desc(fc_strength)) 

distinct_clusters <- tibble(g_clust = data_reclustered$clusters,
                            names = data_reclustered$celltype) |>
  distinct() |>
  arrange(g_clust) |>
  mutate(names = paste(g_clust, names),
         g_clust = paste0("g", g_clust))

```

```{R}
build_chart <- function(your_genes, lab) {
  avg_expr <- AverageExpression(data_reclustered, features = your_genes, group.by = c("clusters", "Treatment"),
                                scale.factor = 100, layer = "scale.data")$integrated
  
  radar_dat <- avg_expr %>%
    as.matrix() %>%
    as.data.frame() %>%
    rownames_to_column("Gene") %>%
    pivot_longer(cols = -1,
                 names_to = c("g_clust", "Treatment"),
                 names_sep = "_",
                 values_to = "value") %>%
    left_join(distinct_clusters, by = "g_clust") %>%
    select(-g_clust) %>%
    pivot_wider(names_from = names) %>%
    mutate(Treatment = recode(Treatment,
                              "control" = "Control",
                              "Dtreatment" = "Dexmethasone",
                              "PDtreatment" = "Paclitaxel + Dexmethasone",
                              "Ptreatment" = "Paclitaxel"))
  
  set.seed(51)
  palette <- distinctColorPalette(n_top_genes)
  
  input_dat <- radar_dat %>%
    group_by(Treatment) %>%
    group_split() 
  
  mk_radar_plot <- function(data) {
    this_trt <- unique(data$Treatment)
    data |>
      ggRadar(aes(group = "Gene"), alpha = 0.3) + # alpha for polygon colors intensity
      theme_bw() +
      ggplot2::scale_fill_manual(values = palette) +
      ggplot2::scale_color_manual(values = palette) +
      labs(title = paste("Top", n_top_genes, "enriched genes between Paclitaxel and Control"),
           subtitle = this_trt, caption = lab) +
      theme(axis.text = element_text(size = 7))
    ggsave(file.path(fig_folder, "Radar", paste0(lab, "Radarchart_", stringr::str_remove(this_trt, " \\+ "), ".png")),
           units = "in", width = 7, height = 5)
  }
  
  lapply(input_dat, mk_radar_plot)
}

genes_of_interest |>
  filter(logFC > 0) |>
  slice_head(n = n_top_genes) |>
  pull(gene) |>
  build_chart(lab = "Upregulated")

genes_of_interest |>
  filter(logFC < 0) |>
  slice_head(n = n_top_genes) |>
  pull(gene) |>
  build_chart(lab = "Downregulated")

genes_of_interest |>
  slice_head(n = n_top_genes) |>
  pull(gene) |>
  build_chart(lab = "Most differentially expressed")
```


# Dot plots / heatmaps

```{R}
# DEGs between Paclitaxel and Control
n_top_genes <- 50
DEG_interest <- genes_of_interest %>%
  slice_head(n = n_top_genes) %>%
  pull(gene)

# Get average expression values
avg_expr <- AverageExpression(data_reclustered, features = DEG_interest,
                              group.by = c("clusters", "Treatment"), layer = "scale.data")$integrated

avg_dat <- avg_expr %>%
  as.matrix() %>%
  as.data.frame() %>%
  rownames_to_column("Gene") %>%
  pivot_longer(cols = -1,
               names_to = c("g_clust", "Treatment"),
               names_sep = "_",
               values_to = "value") %>%
  mutate(clusters = as.numeric(stringr::str_remove(g_clust, "g"))) |>
  left_join(distinct_clusters, by = "g_clust") %>%
  select(-g_clust)

# Get pcts
saved_plot <- DotPlot(data_reclustered, features = unique(DEG_interest),
        assay = "integrated",
        cols = palette, scale = TRUE,
        split.by = "Treatment",
        group.by = c("clusters")) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

new_dat <- saved_plot$data %>%
  separate(id, into = c("clusters", "Treatment"), sep = "_") |>
  remove_rownames() |>
  mutate(clusters = as.numeric(clusters), Gene = features.plot) |>
  left_join(avg_dat, by = c("clusters", "Treatment", "Gene"))

ggplot(new_dat, aes(x = names, y = features.plot)) +
  geom_point(aes(size = pct.exp, color = value)) +
  scale_color_viridis_c() +             
  scale_size_continuous(range = c(1, 5)) +
  theme_minimal() +                       
  labs(x = "Cluster", y = "Gene", 
       title = "Dotplot Heatmap of Differentially Expressed Genes",
       subtitle = "Markers between Paclitaxel and Control",
       size = "Pct. Expressed", color = "Avg. Expression") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 5)) +
  coord_flip() +
  facet_grid(facets = "Treatment")
# these are up or downregulated in Paclitaxel

ggsave(file.path(fig_folder, "dotplot.png"), width = 12, units = "in", height = 10)
```
## Only for treatment interesting genes
```{R}

# WHat are the diff expressed genes
n_top_genes <- 75
DEG_interest <- read.csv(file.path(fig_folder, "top_DEG_P_PD.csv")) %>%
  mutate(fc_strength = abs(logFC)) %>%
  arrange(fc_strength) %>%
  slice_head(n = n_top_genes) %>%
  pull(X)

avg_expr <- AverageExpression(data_reclustered, features = DEG_interest,
                              group.by = c("Treatment", "clusters"), layer = "scale.data")$integrated

avg_dat <- avg_expr %>%
  as.matrix() %>%
  as.data.frame() %>%
  rownames_to_column("Gene") %>%
  pivot_longer(cols = -1,
               names_to = c("Treatment", "g_clust"),
               names_sep = "_",
               values_to = "value") %>%
  mutate(clusters = as.numeric(g_clust),
    g_clust = paste0("g", g_clust)) |>
  left_join(distinct_clusters, by = "g_clust") %>%
  select(-g_clust)
  ##mutate(Treatment = recode(Treatment,
  ##                          "control" = "Control",
  #                          "Dtreatment" = "Dexmethasone",
  #                          "PDtreatment" = "Paclitaxel + Dexmethasone",
  #                          "Ptreatment" = "Paclitaxel"))


# Get pcts
saved_plot <- DotPlot(data_reclustered, features = unique(DEG_interest),
        assay = "integrated",
        cols = palette, scale = TRUE,
        split.by = "Treatment",
        group.by = c("clusters")) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

new_dat <- saved_plot$data %>%
  separate(id, into = c("clusters", "Treatment"), sep = "_") |>
  filter(features.plot %in% DEG_interest) |>
  remove_rownames() |>
  mutate(clusters = as.numeric(clusters), Gene = features.plot) |>
  left_join(avg_dat, by = c("clusters", "Treatment", "Gene"))

ggplot(new_dat, aes(x = names, y = features.plot)) +
  geom_point(aes(size = pct.exp, color = value)) +
  scale_color_viridis_c() +             
  scale_size_continuous(range = c(1, 5)) +
  theme_minimal() +                       
  labs(x = "Cluster", y = "Gene", 
       title = "Dotplot Heatmap of Differentially Expressed Genes",
       subtitle = "Markers between P and P+D",
       size = "Pct. Expressed", color = "Avg. Expression") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 5)) +
  coord_flip() +
  facet_grid(facets = "Treatment")
# these are up or downregulated in Paclitaxel

ggsave(file.path(fig_folder, "dotplot_trt_deg.png"), width = 12, units = "in", height = 10)
```
## Another heatmap
High = yellow, low = purple
```{R}
top10 <- all_markers %>%
  mutate(fc_strength = abs(avg_log2FC)) %>%
  arrange(fc_strength) %>%
  group_by(cluster) |>
  slice_head(n = 10) %>%
  select(gene, cluster)

Seurat::DoHeatmap(subset(data_reclustered, downsample = 1000), features = top10$gene,
                  group.by = "ident")

ggsave(file.path(fig_folder, "heatmap_rec.png"), units = "in", width = 5, height = 6)

```

## RAF1
```{R}
seurat_obj <- ScaleData(data_reclustered, assay = "integrated")
avg_expr <- AverageExpression(seurat_obj, features = "Raf1",
                              group.by = c("SingleR.pruned", "Treatment"),
                              scale.factor = 100)$integrated
radar_dat <- avg_expr %>%
  as.matrix() %>%
  as.data.frame() %>%
  pivot_longer(cols = everything(),
               names_to = c("Celltype", "Treatment"),
               names_sep = "_",
               values_to = "value") %>%
  mutate(Treatment = recode(Treatment,
                            "control" = "Control",
                            "Dtreatment" = "Dexmethasone",
                            "PDtreatment" = "Paclitaxel + Dexmethasone",
                            "Ptreatment" = "Paclitaxel")) |>
  group_by(Treatment)

ggplot(radar_dat, aes(x = Celltype, y = value, fill = Treatment)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(title = "Expression of RAF1 averaged across cells",
       ylab = "Average expression level") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0, size = 5))

ggsave(file.path(fig_folder, "RAF1.png"), width = 8, height = 5)
```

# Isolate the top biomarkers

## looking at the celldex set
```{R}
mouse_ref <- celldex::MouseRNAseqData()

all_dat <- SummarizedExperiment::assay(mouse_ref, assay = "logcounts") |> as.data.frame() |>
  rowMeans()

ind <- mouse_ref$label.fine == "B cells"


b_cell_markers <- all_dat[ind] |> sort(decreasing = TRUE)

# Print the B cell markers
print(b_cell_markers)
```

## Dot plot for differential B cells
```{R}
# WHat are the diff expressed genes
n_top_genes <- 5
DEG_interest <- c("Cd74", "H2-Aa", "Actb", "Eef1a1", "Eef2", "Rpl4", "H2-Ab1", "Coro1a", "Tmsb4x", "Ly6e", "H2-Eb1", "Rplp0", "Serinc3", "Gnb2l1", "Cd79b", "Malat1", "H2-K1", "Rps3", "Rpl8", "Rpl13a")
  

# Get average expression values
avg_expr_2 <- AverageExpression(data_reclustered, features = DEG_interest,
                              group.by = c("clusters", "Treatment"), layer = "scale.data")$integrated

avg_dat <- avg_expr_2 %>%
  as.matrix() %>%
  as.data.frame() %>%
  rownames_to_column("Gene") %>%
  pivot_longer(cols = -1,
               names_to = c("g_clust", "Treatment"),
               names_sep = "_",
               values_to = "value") %>%
  mutate(clusters = as.numeric(stringr::str_remove(g_clust, "g"))) #|>
  #left_join(distinct_clusters, by = "g_clust") %>%
  #select(-g_clust)
  ##mutate(Treatment = recode(Treatment,
  ##                          "control" = "Control",
  #                          "Dtreatment" = "Dexmethasone",
  #                          "PDtreatment" = "Paclitaxel + Dexmethasone",
  #                          "Ptreatment" = "Paclitaxel"))


# Get pcts
saved_plot <- DotPlot(data_reclustered, features = unique(DEG_interest),
        assay = "integrated",
        cols = palette, scale = TRUE,
        split.by = "Treatment",
        group.by = c("clusters")) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

ggsave(file.path(fig_folder, "dotplot_bcells.png"), saved_plot, width = 12, units = "in", height = 10)
```

# FIGURE 1

## Standardize colors

```{R}
set.seed(22)

data_reclustered$SingleR.pruned <- data_reclustered$SingleR.pruned |> stringr::str_to_sentence()

# Get all unique cell types from the dataset
all_celltypes <- unique(data_reclustered$SingleR.pruned)
palette <- distinctColorPalette(length(all_celltypes))
names(palette) <- all_celltypes
```

## Stacked bar plot with prop of cells by treatment
```{R}
## Order the cell types
order_celltypes <- data_reclustered@meta.data |>
  group_by(SingleR.pruned) |>
  summarise(count = n()) |>
  arrange((count)) |>
  pull(SingleR.pruned) 

plotting_data <- data_reclustered@meta.data |>
  mutate("Cell type" = factor(SingleR.pruned, levels = order_celltypes))

plotting_data |>
  filter(!is.na(SingleR.pruned)) |>
  ggplot2::ggplot(aes(x = Treatment, fill = `Cell type`)) +
    geom_bar(stat = "count") +
    ggplot2::scale_fill_manual(values = palette) +
    ggplot2::theme_bw() +
    ggplot2::xlab("Treatment") +
    ggplot2::ylab("Cell count") +
  ggplot2::coord_flip() +
  ggplot2::labs(title = "Differential cell type composition across treatments") +
   theme(legend.position = "bottom") +
  guides(color = guide_legend(nrow = 4)) 

ggsave(file.path(fig_folder, "celltypes_trt_prop_FIG1.png"), height = 4, width = 9, units = "in")
```


## UMAP with cluster names
```{R}
# UMAP
input_umap_dat <- Embeddings(data_reclustered, reduction = "umap") |>
  as.data.frame() |>
  rownames_to_column() |>
  left_join(rownames_to_column(data_reclustered@meta.data),
            by = "rowname") |>
  dplyr::rename("Cluster type" = "celltype",
                "Cluster id" = "clusters")

# Compute centroids
centroids <- input_umap_dat |>
  group_by(`Cluster id`, `Cluster type`) |>
  summarise(umap_1 = mean(umap_1), umap_2 = mean(umap_2))

# UMAP plot with centroid labels
in_umap_shuff <- input_umap_dat[sample(nrow(input_umap_dat)), ]
  
in_umap_shuff |>
  mutate("Cell type" = factor(stringr::str_to_sentence(fctr_celltype), levels = order_celltypes)) |>
  ggplot2::ggplot(aes(x = umap_1, y = umap_2,
                      colour = `Cell type`)) +
  geom_point(alpha = 0.4, size = 0.4) +
  geom_text_repel(data = centroids,
                  aes(x = umap_1 + 0.5, y = umap_2 + 0.5, label = `Cluster id`),
                  size = 3, fontface = "bold", color = "black",
                  box.padding = 0.3, point.padding = 0.2,
                  segment.color = NA, inherit.aes = FALSE) +
  labs(x = "UMAP Axis 1", y = "UMAP Axis 2",
       title = "UMAP visualization of cell type diversity across treatments") +
  scale_color_manual(values = palette) +
  theme_classic() +
  guides(color = guide_legend(override.aes = list(alpha = 1, size = 5)))

ggsave(file.path(fig_folder, "UMAP_across_trt_FIG1.png"),
       width = 6, height = 4.5)
```

### Supplementary table 2: how many cells by cluster

```{R}
# Cluster | Cell type | number | percent of total cells in cluster

s2_tab <- in_umap_shuff |>
  mutate("Cell type" = factor(stringr::str_to_sentence(fctr_celltype), levels = order_celltypes)) |>
  group_by(`Cluster id`, `Cell type`) |>
  summarise(cell_count = n(), .groups = "drop_last") |>
  mutate(percent = round((cell_count / sum(cell_count)) * 100, 2))

write_csv(s2_tab, file.path(fig_folder, "s2_tab.csv"))
```

## Barplot of cell types by treatment

```{R}
cluster_matched_cells <- tibble(cluster = Idents(data_reclustered), data_reclustered[["SingleR.pruned"]])

summary_main <- cluster_matched_cells |>
  arrange(cluster) |>
  group_by(cluster, SingleR.pruned) |>
  summarise(num_each = n()) |>
  arrange(cluster, desc(num_each))

 summary_main |>
    dplyr::mutate(proportion = num_each / sum(num_each),
                  "Cell type" = factor(SingleR.pruned, levels = order_celltypes)) |>
    ggplot2::ggplot(aes(x = cluster, y = num_each, fill = `Cell type`)) +
    geom_bar(stat = "identity") +
    ggplot2::scale_fill_manual(values = palette) +
    ggplot2::theme_bw() +
    labs(main = "Cell Type Composition Across Clusters") +
    ggplot2::xlab("Cluster") +
    ggplot2::ylab("Total cell count") +
   theme(legend.position = "bottom") +
   labs(title = "Proportion of cell types by cluster",
        subtitle = "All treatments")

ggsave(file.path(fig_folder, "barplot_main_labels_FIG1.png"),
        width = 9, height = 5, units = "in")

```
### Supplementary table 3: how many cells by treatment

```{R}
# Treatment | Cell type | number | percent of total cells in cluster

s3_tab <- in_umap_shuff |>
  mutate("Cell type" = factor(stringr::str_to_sentence(fctr_celltype), levels = order_celltypes)) |>
  group_by(Treatment, `Cell type`) |>
  summarise(cell_count = n(), .groups = "drop_last") |>
  mutate(percent = round((cell_count / sum(cell_count)) * 100, 2))

write_csv(s3_tab, file.path(fig_folder, "s3_tab.csv"))
```

# Additional Figures (presentation)

## Single gene boxplots

```{R}
# genes <- c("SPP1", "TIMP2") |> str_to_sentence()
# directions <- c("Downregulated", "Downregulated")
# which_trts <- c("control", "Ptreatment")

create_gene_boxplot <- function(genes, directions, which_trts) {
  avg_mat <- AverageExpression(
    data_reclustered,
    features = genes,
    group.by = c("Treatment"),
    assays = "integrated",
    slot = "scale.data",
    normalization.method = "integrated",
    scale.factor = 100,
    layer = "scale.data"
  )$integrated |>
    as.data.frame() |>
    dplyr::select(all_of(which_trts))
  
  df_expr <- avg_mat |>
    rownames_to_column("gene") |>
    pivot_longer(-gene, names_to = "group", values_to = "mean_expression") |>
    mutate(direction = directions[match(gene, genes)])
  
  # Plot
  p <- df_expr |>
    mutate(group = factor(group,
                          levels = c("control", "Ptreatment", "Dtreatment", "PDtreatment"),
                          labels = c("Control", "PTX", "DEX", "PTX + DEX"))) |>
    ggplot(aes(x = gene, y = mean_expression)) +
    geom_bar(stat = "identity", position = position_dodge(),
             fill = "steelblue") +
    facet_wrap(~ group) +
    labs(title = "Gene Expression by Treatment Group",
         y = "Average Scaled Expression",
         x = "Gene") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 20, hjust = 1))
  print(p)
  ggsave(file.path(fig_folder, paste("deg_barplot_calcific.png", sep = "_")),
         p, units = "in", width = 4, height = 4)
}


# Control vs. other treatments
create_gene_boxplot(genes = c("SPP1", "TIMP2") |> str_to_sentence(),
                    directions = c("Downregulated", "Downregulated"),
                    which_trts = c("control", "Ptreatment", "PDtreatment", "Dtreatment"))
```

